codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ServiceGenerator.java

```diff
@@ -0,0 +1,136 @@
+
+/**
+ * Generates a service client and configuration.
+ */
+final class ServiceGenerator implements Runnable {
+
+    public static final String CONFIG_NAME = "Options";
+    public static final String API_OPTIONS_FUNC_NAME = "APIOptionFunc";
+
+    private final GoSettings settings;
+    private final Model model;
+    private final SymbolProvider symbolProvider;
+    private final GoWriter writer;
+    private final ServiceShape service;
+    private final List<GoIntegration> integrations;
+    private final List<RuntimeClientPlugin> runtimePlugins;
+
+    ServiceGenerator(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            ServiceShape service,
+            List<GoIntegration> integrations,
+            List<RuntimeClientPlugin> runtimePlugins
+    ) {
+        this.settings = settings;
+        this.model = model;
+        this.symbolProvider = symbolProvider;
+        this.writer = writer;
+        this.service = service;
+        this.integrations = integrations;
+        this.runtimePlugins = runtimePlugins;
+    }
+
+    @Override
+    public void run() {
+        Symbol serviceSymbol = symbolProvider.toSymbol(service);
+        writer.writeShapeDocs(service);
+        writer.openBlock("type $T struct {", "}", serviceSymbol, () -> {
+            writer.write("options $L", CONFIG_NAME);
+        });
+
+        generateConstructor(serviceSymbol);
+
+        String clientId = CodegenUtils.getDefaultPackageImportName(settings.getModuleName());
+        String clientTitle = service.getTrait(TitleTrait.class).map(StringTrait::getValue).orElse(clientId);
+
+        writer.writeDocs("ServiceID returns the name of the identifier for the service API.");
+        writer.write("func (c $P) ServiceName() string { return $S }", serviceSymbol, clientId);
+
+        writer.writeDocs("ServiceName returns the full service title.");
+        writer.write("func (c $P) ServiceID() string { return $S }", serviceSymbol, clientTitle);
+
+        generateConfig();
+
+        Symbol stackSymbol = SymbolUtils.createPointableSymbolBuilder("Stack", GoDependency.SMITHY_MIDDLEWARE).build();
+        writer.write("type $L func($P) error", API_OPTIONS_FUNC_NAME, stackSymbol);
+    }
+
+    private void generateConstructor(Symbol serviceSymbol) {
+        writer.writeDocs(() -> writer.write(
+                    "New returns an initialized $T based on the functional options.\n"
+                    + "Provide additional functional options to further configure the behavior\n"
+                    + "of the client, such as changing the client's endpoint or adding custom\n"
+                    + "middleware behavior.", serviceSymbol
+        ));
+        writer.openBlock("func New(options $L) ($P, error) {", "}",
+                CONFIG_NAME, serviceSymbol, () -> {
+
+            writer.write("resolvedOptions = options.Copy()").write("");
```

Comment:
Its find to replace the original `options` value`. since we want to ensure nothing later ever refers to it by mistake.

```go
options = options.Copy()
```